migrations/educandu-2022-09-07-02-streamline-nulls-in-plugins.js

```diff
@@ -0,0 +1,227 @@
+    let sectionWasUpdated = false;
+
+    if (!section.content) {
+      return sectionWasUpdated;
+    }
+
+    if (section.type === 'markdown' && 'renderMedia' in section.content) {
+      // Delete superfluous property
+      delete section.content.renderMedia;
+      sectionWasUpdated = true;
+    }
+
+    if (section.type === 'annotation' && 'renderMedia' in section.content) {
+      // Delete superfluous property
+      delete section.content.renderMedia;
+      sectionWasUpdated = true;
+    }
+
+    if (section.type === 'table' && 'renderMedia' in section.content) {
+      // Delete superfluous property
+      delete section.content.renderMedia;
+      sectionWasUpdated = true;
+    }
+
+    if (section.type === 'interactive-media' && 'range' in section.content) {
+      // Delete wrongly named property
+      delete section.content.range;
+      sectionWasUpdated = true;
+    }
+
+    if (section.type === 'image') {
+      if (section.content.copyrightNotice === null) {
+        // Convert null -> empty string
+        section.content.copyrightNotice = '';
+        sectionWasUpdated = true;
+      }
+      if (section.content.effect?.type === EFFECT_TYPE.clip && 'copyrightNotice' in section.content.effect) {
+        // Delete superfluous property
+        delete section.content.effect.copyrightNotice;
+        sectionWasUpdated = true;
+      }
+    }
+
+    if (section.type === 'audio') {
+      if (section.content.sourceUrl === null) {
+        // Convert null -> empty string
+        section.content.sourceUrl = '';
+        sectionWasUpdated = true;
+      }
+      if (section.content.copyrightNotice === null) {
+        // Convert null -> empty string
+        section.content.copyrightNotice = '';
+        sectionWasUpdated = true;
+      }
+    }
+
+    if (section.type === 'video') {
+      if (section.content.sourceUrl === null) {
+        // Convert null -> empty string
+        section.content.sourceUrl = '';
+        sectionWasUpdated = true;
+      }
+      if (section.content.copyrightNotice === null) {
+        // Convert null -> empty string
+        section.content.copyrightNotice = '';
+        sectionWasUpdated = true;
+      }
+    }
+
+    if (section.type === 'catalog') {
+      for (const item of section.content.items) {
+        // Ensure property
+        if (!('documentId' in item.link)) {
+          item.link.documentId = null;
+          sectionWasUpdated = true;
+        }
+        // Convert empty string -> null
+        if (item.link.documentId === '') {
+          item.link.documentId = null;
+          sectionWasUpdated = true;
+        }
+      }
+    }
+
+    if (section.type === 'ear-training') {
+      for (const test of section.content.tests) {
+        // Migrate leftover older data structure
+        if ('startAbcCode' in test) {
+          test.mode = TEST_MODE.abcCode;
+          test.questionAbcCode = test.startAbcCode;
+          test.answerAbcCode = test.fullAbcCode;
```

Comment:
Interesting 😳 Is it possible that the migration was already done, but without the cleanup?
If so, we probably should prioritize the existing data (`test.questionAbcCode = test.questionAbcCode || test.startAbcCode` for safety?)

Reply:
The data I checked on prod had both the old and new props, so I didn't want to "revive" the old values from their grave (since nobody has so far complained that they have been lost obviously some while ago)

Reply:
Ok, then R.I.P 😄

Reply:
I checked, and now I remember the reason for the code. All sections that have `startAbcCode` set, do not have any of the new properties (meaning they were never migrated), so the migration script will actually fix this older issue. There is no need for the `test.questionAbcCode || test.startAbcCode` fallback.

Reply:
I see. Yet another good find then. Glad this whole correction will happen now with this PR.